client-go/workqueue: Drain work queue on shutdown

alexanderConstantinescu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind api-change

#### What this PR does / why we need it:

This PR implements a drained shutdown of the workqueue, in line with what the doc bloc mention for `Shutdown` in both `delaying_queue.go` and `queue.go`. This drained shutdown was not the case today (i.e: the doc bloc were wrong). The underlying functional reason for this PR is that it would be better if controllers finished executing their `syncHandler` before shutting down completely, some controllers might depend (or are at least better off) being able to terminate that sync term, processing the result and updating whatever data store they depend on, before terminating completely.

This PR assumes that whatever is put on the queue will also terminate at one point (I am not sure if this is a valid assumption), i.e: that no `syncHandler` will indefinitely execute. It thus does not implement a timeout when verifying if the queue is still processing. 

Moreover, this introduces an API change to the client-go's workqueue library. I am thus not sure if this requires additional work besides this PR?

Assigning to the people whom - it seems - have most frequently come in contact with the workqueue. 

/assign @deads2k @lavalamp 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added a `Processing` condition for the workqueue API
Changed `Shutdown` for the workqueue API to wait until the work queue finishes processing all in-flight items.  
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

alexanderConstantinescu committed:
[queue] Implement `ShutDownWithDrain` allowing the queue to drain when shutting down

Signed-off-by: Alexander Constantinescu <aconstan@redhat.com>

alexanderConstantinescu commented:
Thinking about this further, I am not sure if this is considered a bug: that would depend what Kubernetes defines is its contract to the user. If the doc bloc for public API functions is considered a contract, then this could qualify as a bug, since the implementation does not up-hold its publicly stated contract.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:
```diff
@@ -147,8 +147,13 @@ func (pq waitForPriorityQueue) Peek() interface{} {
+// queue will also finish processing at some point in time, hence: no need to
+// use a timeout. This method may be invoked more than once.
 func (q *delayingType) ShutDown() {
+	for q.Processing() {
```
This doesn't seem like the right way to do this, what is to prevent people from adding more items?

I don't know how the delaying queue works (I wrote the non-delaying one) but generally a shut down should make sure that callers attempting to add new items to the queue get an error or at least don't actually add something to the queue.

alexanderConstantinescu replied:
Indeed. I've updated my PR to have the verification of the `processing` length happen after we set `shuttingDown`, as such the queue will stop new items from being accepted but `ShutDown` will still wait until all items have been processed before returning - and thus have the controller process wait for the `syncHandler` to finish for all items before terminating.

caesarxuchao commented:
/triage accepted
/cc @yliaog

alexanderConstantinescu commented:
/remove-kind api-change 

since the new version of this patch does not expose `Processing` through the interface and has the API remain the same.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -181,10 +185,24 @@ func (q *Type) Done(item interface{}) {
-// ShutDown will cause q to ignore all new items added to it. As soon as the
+// Processing indicates if there are still items on the workqueue being processed.
+// It's used to drain the workqueue on an eventual shutdown
+func (q *Type) Processing() bool {
```
I don't think this should be a public function, for any other use it basically isn't threadsafe, since a new item could get added after the unlock before the function return.

alexanderConstantinescu replied:
Yes, I agree. Made it private

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -181,10 +185,24 @@ func (q *Type) Done(item interface{}) {
+	return q.processing.len() != 0
+}
+
 // worker goroutines have drained the existing items in the queue, they will be
```
part of this comment got deleted accidentally

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -159,3 +159,72 @@ func TestReinsert(t *testing.T) {
+	firstItem, _ := q.Get()
+	secondItem, _ := q.Get()
+
+	finishedShutdownTest := finishedShutdownTest{}
```
why not use a sync.WaitGroup?

alexanderConstantinescu replied:
The goal I was trying to achieve by assigning `finishedShutdownTest.finished` was to verify that `ShutDown` never returned during the test execution until the very end when we've marked all items as `Done`. It's true that using a `WaitGroup` achieves the same goal more robustly, since if we `Wait` before the last `Done` is called, we would fail the test. Conversely, if we're missing a `Done` for any item on the queue: we also fail the test and time out.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -159,3 +159,72 @@ func TestReinsert(t *testing.T) {
+		finishedShutdownTest.setFinished()
+	}()
+
+	time.Sleep(20 * time.Millisecond)
```
This could be racy in busy CI environments. You'll have to do something more obscure, like adding items until that fails and then verifying that `ShuttingDown()` returns true.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -159,3 +159,72 @@ func TestReinsert(t *testing.T) {
+	}
+	// Sleep a maximum of 1 second since that's the interval used to check the
+	// queue size
+	time.Sleep(1 * time.Second)
```
This will be racy. Turn `finishedShutdownTest` into a `sync.WaitGroup`, call `.Wait()` here, time how long it takes if you really want to, but leave at least a 3x margin if you want to fail if it takes too long (I don't think that's necessary).

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -159,3 +159,72 @@ func TestReinsert(t *testing.T) {
+	}
+}
+
+type finishedShutdownTest struct {
```
(You should use wait group in this case. But if you ever want to do something like this in the future, I recommend https://golang.org/pkg/sync/atomic/#Value instead)

alexanderConstantinescu replied:
Cheers, and thanks for all the pointers and review on this!

deads2k commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -181,10 +185,25 @@ func (q *Type) Done(item interface{}) {
 // ShutDown will cause q to ignore all new items added to it. As soon as the
 // worker goroutines have drained the existing items in the queue, they will be
 // instructed to exit.
 func (q *Type) ShutDown() {
```
This is a reasonable idea on the *first* sig-int.  But on the second interrupt or on sig-term we should directly shutdown directly.  Do that, I think you need to pass a `secondInterruptCh`.

alexanderConstantinescu replied:
I've implemented a signal notifier which will set `forceShutdown` and interrupt the drain. I am however not sure I like it. Kubernetes controllers seems to be very opinionated when it comes to handling multiple signal interrupts, and it seems [like best practices](https://github.com/kubernetes/sample-controller/blob/master/pkg/signals/signal.go#L36-L39) even exit the process completely. Also, given that signals aren't uniform cross OS platforms I needed to add dedicated OS parts to this PR to handle this (which I am not sure the queue should / can handle).

lavalamp replied:
This package should not bake in the signal handling; that's up to the caller, there are other reasons to want to shut down.

I think what David is asking for, is something more like a second ShutDownNoWaiting method which doesn't block for the queue to clear, and ideally makes any callers blocking on ShutDown return immediately also.

alexanderConstantinescu replied:
I think that makes more sense. I've implemented `ShutDownNoDrain` now which will not wait for the queue to drain when invoked, and also immediately release any callers waiting on `ShutDown`. Callers can thus implement their own signal handler and invoke `ShutDownNoDrain` in whatever situation they see fit as to force the shut down to finish immediately.

deads2k replied:
> I haven't reviewed the tests. The interface looks reasonable to me now. Perhaps @deads2k can confirm.

I think I'm ok interface-wise, but we have now changed the contract without changing the method signature.  This call used to cause all .Gets to return as shutdown immediately and the method did not block. Now the method blocks.

This means we have to update our callers, so that in our cases (KCM, KS, KAS, etc) we have the first ctrl+c call shutdown (as we do today I think), and a second ctrl+c call a `ShutdownNoDrain`, right?

alexanderConstantinescu replied:
> but we have now changed the contract without changing the method signature

I don't fully agree. We have not changed the contract, the contract said that it will drain the queue on shut down and this is not what it does today. What has changed is the implementation: which now does what the contract says. 

> This means we have to update our callers

Not necessarily. If a caller could accept a false contract, it should be able to accept a true one. Callers today usually forcefully end up doing `oc.Exit(1)` on the second ctrl+c call, this can continue being the case if callers do not want to use `ShutDownNoDrain`. The only real change this patch implements, for what concerns existing callers, is: allowing the queue to finish cleaning up after first ctrl+c and until the second ctrl+c

lavalamp replied:
> we have now changed the contract without changing the method signature

Sorry, I missed this, I thought it was net-new; I agree we should introduce a ShutdownWithDrain rather than a ShutdownNoDrain.

> If a caller could accept a false contract, it should be able to accept a true one.

Tragically, this is emphatically not true in software :( :( :(

alexanderConstantinescu replied:
> I agree we should introduce a ShutdownWithDrain rather than a ShutdownNoDrain.

OK, I've updated the PR and implemented `ShutDownWithDrain`, leaving `ShutDown` as is (i.e: not blocking and not draining). It is safe to call `ShutDown` after `ShutDownWithDrain` as to force the queue to shut down immediately on the callers decision (ex: when receiving a dedicated OS signal, SIGTERM/SIGINT)

deads2k commented:
/ok-to-test

alexanderConstantinescu commented:
Update: I've implemented a "force shutdown" mechanism as requested on the latest comment. I've also added the commit `use cond.Wait to wait for the queue draining instead of sleeping` which should improve things, since we can remove the sleep and more accurately finish draining the queue.

alexanderConstantinescu commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -198,3 +200,81 @@ func TestFinishShutDownUntilAllItemsAreDone(t *testing.T) {
+
+	// Only send os.Interrupt here, as to make sure that the tests can be
+	// executed OS agnostically
+	if err := proc.Signal(os.Interrupt); err != nil {
```
From the doc of `os.Interrupt`
```
On Windows, sending os.Interrupt to a process with os.Process.Signal is not implemented; it will return an error instead of sending a signal.
```
question: does Kubernetes run its tests on Windows? This test will not pass in that case.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -61,7 +63,7 @@ func newQueue(c clock.Clock, metrics queueMetrics, updatePeriod time.Duration) *
 		go t.updateUnfinishedWorkLoop()
 	}
-
+	t.setupSignalHandler()
```
It needs to be up to the caller to do this, IMO.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -175,22 +181,69 @@ func (q *Type) Done(item interface{}) {
 	q.metrics.done(item)
 
 	q.processing.delete(item)
+	if q.processing.len() == 0 {
```
I feel like this could be in an else clause on the following `if` statement?

alexanderConstantinescu replied:
It could. The following:
```go
if q.dirty.has(item) {
        q.queue = append(q.queue, item)
	q.cond.Signal()
} else if q.processing.len() == 0 {
        q.cond.Signal()
}
```
would lead to less `Signal` invocations than the current version

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -175,22 +181,69 @@ func (q *Type) Done(item interface{}) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	q.drain = false
+	q.cond.Broadcast()
```
I'm not sure both `noDrain` and `shutdown` need to broadcast?

alexanderConstantinescu replied:
Indeed, given that they are called successively it is fine to only have `shutdown` broadcast.

alexanderConstantinescu commented:
/test pull-kubernetes-unit

deads2k commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -178,19 +185,64 @@ func (q *Type) Done(item interface{}) {
 		q.queue = append(q.queue, item)
 		q.cond.Signal()
+	} else if q.processing.len() == 0 {
+		q.cond.Signal()
```
Trying to reason through this.  In a typical usage we have multiple go routines doing `.Get` which has a `.wait` today.  The `.signal` from an Add and here generally ensure that when an item is added to the queue we have a matching `.signal` for every `.wait`.  This is leveraged in `.Get` so that a signal with an empty queue indicates that there is no more work to do.

This new signal appears to used to indicate that there is no processing.  Now the `.Get` calls will continue returning items until the queue is empty, then they `.wait`.

If we have two workers doing Gets and one call to Shudown and item/A  I think we can a situation like....
1. Get/1 receives item/A
2. Get/2 enters and waits
3. Shutdown is called and enters a wait
4. we now have two waiters
5. Done for item/A is called and signals
6. Get/2 is signaled from Done and exits as shutting down
7. Get/1 enters and exits as shutting down
8. Shutdown is never signaled

Is that a flow that can happen in this scenario?

Do you actually need the if-block in here to test for shutdown, number in processing, the number in queue, the number in dirty and broadcast?

If I'm correct, I'd like to see a test trying to force this race and fix it.  If I'm incorrect, but the story was plausible to either @alexanderConstantinescu or @lavalamp I think we also need this test.  If I'm incorrect, but the story was not plausible, please help me see the flaw.

alexanderConstantinescu replied:
> the story was plausible to either @alexanderConstantinescu or @lavalamp I think we also need this test.

The story is plausible in my opinion. I will write a test for it and verify. If this is true, I suspect doing 

```go
else if q.processing.len() == 0 {		
        q.cond.Broadcast()
}
```

should fix it, as that would lead to:

5. Done for item/A is called and signals
6. Get/2 and Shutdown is signaled from Done and exits as shutting down
7. Get/1 enters and exits as shutting down

alexanderConstantinescu replied:
After having thought about this some more and written a small test to verify: I've realized that the scenario presented is not completely correct. The following is the actual chain of events:

1. Get/1 receives item/A
2. Get/2 enters and waits
3. Shutdown is called, broadcasts `shuttingDown` and enters a wait
4. Get/2 is signaled from `shuttingDown` and exits as shutting down
4. we now have one waiter
5. Done for item/A is called and signals
6. Shutdown is signaled from Done and exists

Given that `ShutDown` always broadcasts to all `Get`s that they should finish waiting (so, either process one of the remaining items on the queue or just exit if there are none), and also broadcasts to all `Add`s to stop accepting items: we should be fine.

deads2k commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -178,19 +185,64 @@ func (q *Type) Done(item interface{}) {
 // instructed to exit.
 func (q *Type) ShutDown() {
+	q.shutdown()
+	for q.isProcessing() && q.shouldDrain() {
```
What prevents us from being in a state where we are not processing, but still have items in our queue?

alexanderConstantinescu replied:
Nothing. One example of this is what I mentioned in the description to this PR:

> This PR assumes that whatever is put on the queue will also terminate at one point (I am not sure if this is a valid assumption), i.e: that no syncHandler will indefinitely execute

And if there's a caller (controller) for which that statement does not hold, then we block on `ShutDown` until the second ctrl+c.  

I think for what concerns the logic in `queue.go` : there is no condition I could think of that would lead us to not process and still have items in our queue.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -159,3 +159,106 @@ func TestReinsert(t *testing.T) {
+
+	// We know we are shutting down now and cannot add new items, so mark all
+	// items we added until we started shutting down, as done
+	for i := 0; i <= deltaItems; i++ {
```
I think there is an off by 1 error here.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -159,3 +159,106 @@ func TestReinsert(t *testing.T) {
+		q.ShutDownWithDrain()
+	}()
+
+	// Add as many items as possible until the queue is marked "ShuttingDown"
```
State the purpose of this? Is it to exercise a race condition? If so it might be good to have it run this test many times.

alexanderConstantinescu replied:
The purpose of this was to ensure a sequence of events where `ShutDownWithDrain` is called before we start marking items as done - thus "simulating" that we drain the queue / wait for all items to finish processing. I realized that the way of doing this was a bit contrived given my initial goal and I have now re-written it, removing `deltaItems`: which should make the purpose clearer.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue_test.go:
```diff
@@ -159,3 +159,106 @@ func TestReinsert(t *testing.T) {
+
+	// We know we are shutting down now and cannot add new items, so mark all
+	// items we added until we started shutting down, as done
+	for i := 0; i <= deltaItems; i++ {
```
same comments as on an earlier test.

lavalamp commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -178,13 +184,58 @@ func (q *Type) Done(item interface{}) {
+	q.shutdown()
+}
+
+// ShutDownWithDrain will cause q to ignore all new items added to it. As soon
```
It's important to clarify here: is "draining" about processing every item currently in the queue (including dirty items that will be re-added when a worker exits), or is it about getting every worker goroutine to exit?

alexanderConstantinescu replied:
OK, "draining" is about processing every item currently in the queue. I have update the doc bloc with (what I hope is) a better wording for this.

lavalamp replied:
Great -- in that case, you need to exercise the case where an item is added while it's being processed, I'm pretty sure such items get left in the queue right now.

alexanderConstantinescu replied:
Isn't this what `TestAddWhileProcessing` does? I've update https://github.com/kubernetes/kubernetes/blob/525a2032a4ed5eab44bcbb5f356ebf151c47fa46/staging/src/k8s.io/client-go/util/workqueue/queue_test.go#L113 to use `ShutDownWithDrain` and the test passes. Either that test doesn't test what you want me to test, or there's no problem?

lavalamp replied:
Did you verify the queue was empty when ShutDownWithDrain exited?

I'd prefer adding a very short test that exercised the scenario deterministically.

alexanderConstantinescu replied:
> Did you verify the queue was empty when ShutDownWithDrain exited?

Yes

> I'd prefer adding a very short test that exercised the scenario deterministically.

I've modified both `TestAddWhileProcessing` and `TestBasic` to include both shut down mechanisms to the tests and added a verification on the queue length for both tests and shut downs. I did this to `TestBasic` too, for completeness sake. Let me know if you think this is sufficient / valid.

lavalamp commented:
OK I wrote a test for the scenario I was worried about and it passes, so great. Can you add this test? Perhaps it is already covered but I don't see it.

```go
func TestQueueDrainageUsingShutDownWithDrainWithDirtyItem(t *testing.T) {
        q := workqueue.New()

        q.Add("foo")
        gotten, _ := q.Get()
        q.Add("foo")

        finishedWG := sync.WaitGroup{}
        finishedWG.Add(1)
        go func() {
                defer finishedWG.Done()
                q.ShutDownWithDrain()
        }()

        // Ensure that ShutDownWithDrain has started and is blocked.
        shuttingDown := false
        for !shuttingDown {
                _, shuttingDown = q.Get()
        }

        // Finish "working".
        q.Done(gotten)

        // `shuttingDown` becomes false because Done caused an item to go back into
        // the queue.
        again, shuttingDown := q.Get()
        if shuttingDown {
                t.Fatalf("should not have been done")
        }
        q.Done(again)

        // Now we are really done.
        _, shuttingDown = q.Get()
        if !shuttingDown {
                t.Fatalf("should have been done")
        }

        finishedWG.Wait()
}
```

...unfortunately I also ran with -count to look for flakes and if you pass something like 40 it usually hangs. So there is some rare lock up.

lavalamp commented:
Stack trace from a frozen run:

```
panic: test timed out after 10m0s
                                                                                                         
goroutine 161 [running]:                                                                                                                                                                                           
testing.(*M).startAlarm.func1()                                                                          
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1701 +0xe5                                                                                                                           
created by time.goFunc
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/time/sleep.go:180 +0x45
                                                                                                         
goroutine 1 [chan receive]:                                                                                                                                                                                        
testing.(*T).Run(0xc000001c80, 0x6a3c45, 0x16, 0x6b2d40, 0x490f01)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1240 +0x2da                                                                                                                          
testing.runTests.func1(0xc000102900)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1512 +0x78
testing.tRunner(0xc000102900, 0xc000151d68)                                                              
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1194 +0xef                                                                                                                           
testing.runTests(0xc0001380c0, 0x84e780, 0x14, 0x14, 0xc03dc8a4d8509dbf, 0x8bb2d5261e, 0x853820, 0x203000)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1510 +0x2fe                                                                                                                          
testing.(*M).Run(0xc0001c2000, 0x0)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1418 +0x1eb
k8s.io/client-go/util/workqueue.TestMain(0xc0001c2000)                   
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/main_test.go:28 +0x93       
main.main()                                                                                              
        _testmain.go:89 +0x165                                                                                                                                                                                     

goroutine 19 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x853980)                                                         
        /usr/local/google/home/dbsmith/.gvm/pkgsets/go1.16/global/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1169 +0x8b     
created by k8s.io/klog/v2.init.0                                                                         
        /usr/local/google/home/dbsmith/.gvm/pkgsets/go1.16/global/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:420 +0xdf     

goroutine 10 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc0000122d0, 0xc000000002)                  
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/runtime/sema.go:513 +0xf8                                                                                                                               
sync.(*Cond).Wait(0xc0000122c0)                                                                          
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/sync/cond.go:56 +0x99                                                                                                                                   
k8s.io/client-go/util/workqueue.(*Type).wait(0xc000030a20)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:224 +0x7a
k8s.io/client-go/util/workqueue.(*Type).ShutDownWithDrain(0xc000030a20)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:209 +0x50
k8s.io/client-go/util/workqueue_test.TestAddWhileProcessing(0xc000001c80)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue_test.go:143 +0x1ac
testing.tRunner(0xc000001c80, 0x6b2d40)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1239 +0x2b3

goroutine 25 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00012a480)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 31 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00012ae40)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 9 [chan receive]:
k8s.io/client-go/util/workqueue.(*Type).updateUnfinishedWorkLoop(0xc0000303c0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:256 +0xac
created by k8s.io/client-go/util/workqueue.newQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:63 +0x154

goroutine 36 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00020e4e0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 52 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00012b6e0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 99 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00020e8a0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185
FAIL    k8s.io/client-go/util/workqueue 600.032s
```

alexanderConstantinescu commented:
> ...unfortunately I also ran with -count to look for flakes and if you pass something like 40 it usually hangs. So there is some rare lock up.

Indeed. It's now been fixed.

Background: I modified test case: `TestAddWhileProcessing` to test the scenario for both shut down mechanism, and it was failing sometimes for `ShutDownWithDrain`. The reason as to why it was failing was because there was a slight window in https://github.com/kubernetes/kubernetes/blob/dcec2352c0e0b56854e37d375cc3fac5718e18c5/staging/src/k8s.io/client-go/util/workqueue/queue.go#L208-L210 which allowed for the following condition to occur. 

1. The last processing item is still in the `processing` set, hence `q.isProcessing` returns true
2. The worker go-routine finishes the item up and calls `Done` on it
3. `q.wait` starts waiting for any item to finish up and being called `Done` on, but since the queue now is empty: that never happens.

A very subtle lock up, as mentioned, since the window allowing for this to happen is needless to say: small. 

I've fixed this by now making sure that we do not wait if the queue is empty. 

The tests seem good now:

```
$ go test -timeout 180s k8s.io/client-go/util/workqueue -count 100 -race 
ok      k8s.io/client-go/util/workqueue 107.167s
```

pacoxu commented:
/kind feature

lavalamp commented:
/approve
/lgtm

OK, I believe the tests are good now, I think we caught all problems.
